### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np
# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_df = pd.read_csv(file_to_load)
purchase_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [2]:
# Calculating the total number of players (excluding duplicates)
total_players = len(purchase_df["SN"].unique())
total_players

# Create data frame  
total_players_df = pd.DataFrame({
    "Total Players": [total_players]
})
total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
# Calculate number of unique items 
unique_items = len(purchase_df["Item ID"].unique())

# Calculate Average Purchase Price 
avg_purchase_price = sum(purchase_df["Price"]) / len(purchase_df["Price"])

# Calculate number of purchases 
num_purchases = len(purchase_df["Purchase ID"].unique())

# Calculate Total Revenue
total_revenue = sum(purchase_df["Price"])


In [4]:
# Create summary table for purchase analysis
summary_purchase_analysis = pd.DataFrame({
    "Number of Unique Items": [unique_items],
    "Average Price": [avg_purchase_price],
    "Number of Purchases": [num_purchases],
    "Total Revenue": [total_revenue]
    
})


In [5]:
# Formatting 
summary_purchase_analysis["Average Price"] = summary_purchase_analysis["Average Price"].map("${:.2f}".format)
summary_purchase_analysis["Total Revenue"] = summary_purchase_analysis["Total Revenue"].map("${:,.2f}".format)
summary_purchase_analysis

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [6]:
# Identify the different genders 
gender_list = list(purchase_df["Gender"].unique())

# Filter dataset to only male
male_data_df = purchase_df.loc[purchase_df["Gender"] == "Male", :]
male_data_df

# Calculate percentage and count of Male Players 
male_count = len(male_data_df["SN"].unique())
male_percentage = (male_count / total_players) * 100

# Filter dataset to only female
female_data_df = purchase_df.loc[purchase_df["Gender"] == "Female", :]
female_data_df

# Calculate percentage and count of Female Players 
female_count = len(female_data_df["SN"].unique())
female_percentage = (female_count / total_players) * 100

# Filter dataset to only Other / Non-Disclosed
other_data_df = purchase_df.loc[purchase_df["Gender"] == "Other / Non-Disclosed", :]
other_data_df

# Calculate percentage and count of Other / Non-Disclosed 
other_count = len(other_data_df["SN"].unique())
other_percentage = (other_count / total_players) * 100

In [7]:
# Create summary table for gender demographics
summary_gender_demo = pd.DataFrame({
    "Gender": ["Male","Female","Other / Non-Disclosed"],
    "Total Count": [male_count, female_count, other_count],
    "Percentage of Players": [male_percentage, female_percentage, other_percentage] 
})



In [8]:
# Formatting 
summary_gender_demo["Percentage of Players"] = summary_gender_demo["Percentage of Players"].map("{:.2f}%".format)
summary_gender_demo

Gender  Total Count Percentage of Players
0                   Male          484                84.03%
1                 Female           81                14.06%
2  Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [9]:
# Purchase Count by Gender 
f_purchase_count = female_data_df["Purchase ID"].count()
m_purchase_count = male_data_df["Purchase ID"].count()
o_purchase_count = other_data_df["Purchase ID"].count()

# Average Purchase Price by Gender 
f_avg_price = female_data_df["Price"].mean()
m_avg_price = male_data_df["Price"].mean()
o_avg_price = other_data_df["Price"].mean()

# Total Purchase Value by Gender 
f_total_purchase = female_data_df["Price"].sum() 
m_total_purchase = male_data_df["Price"].sum() 
o_total_purchase = other_data_df["Price"].sum()

# Average Total Purchase per Person by Gender 
f_avg_purchase = f_total_purchase / female_count 
m_avg_purchase = m_total_purchase / male_count
o_avg_purchase = o_total_purchase / other_count


In [10]:
# Create summary table for purchasing analysis by gender 
summary_purchase_analysis_gender = pd.DataFrame({
    "Gender": ["Female","Male","Other / Non-Disclosed"],
    "Purchase Count": [f_purchase_count, m_purchase_count , o_purchase_count],
    "Average Purchase Price": [f_avg_price, m_avg_price, o_avg_price],
    "Total Purchase Value": [f_total_purchase, m_total_purchase, o_total_purchase],
    "Avg Total Purchase Value per Person": [f_avg_purchase, m_avg_purchase, o_avg_purchase]
})


In [11]:
# Formatting 
summary_purchase_analysis_gender["Average Purchase Price"] = summary_purchase_analysis_gender["Average Purchase Price"].map("${:.2f}".format)
summary_purchase_analysis_gender["Total Purchase Value"] = summary_purchase_analysis_gender["Total Purchase Value"].map("${:,.2f}".format)
summary_purchase_analysis_gender["Avg Total Purchase Value per Person"] = summary_purchase_analysis_gender["Avg Total Purchase Value per Person"].map("${:.2f}".format)
summary_purchase_analysis_gender

Gender  Purchase Count Average Purchase Price  \
0                 Female             113                  $3.20   
1                   Male             652                  $3.02   
2  Other / Non-Disclosed              15                  $3.35   

  Total Purchase Value Avg Total Purchase Value per Person  
0              $361.94                               $4.47  
1            $1,967.64                               $4.07  
2               $50.19                               $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [12]:
# Caluclate the min and max ages in dataset
min_age = purchase_df["Age"].min()
max_age = purchase_df["Age"].max()

# Identify the list of ages 
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]

# Create the names for the bins 
group_names = ["<10", "10-14", "15-19","20-24","25-29","30-34","35-39","40+"]

# Categorize existing players using age bin
purchase_df["Age Group"] = pd.cut(purchase_df["Age"], bins, labels = group_names)

# Group the purchase_df dataframe by the age demographics and 
# count number of *unique* players that fall within the bins 
age_group_df = purchase_df.groupby(["Age Group"])
age_count = purchase_df.groupby(["Age Group"])["SN"].nunique()

# Calculate percentages by age group
age_percent = (age_count / total_players) * 100

In [13]:
# Create summary table for age demographics by gender 
summary_age_demo_gender = pd.DataFrame({
    "Total Count": age_count,
    "Percentage of Players": age_percent
})


In [14]:
# Formatting 
summary_age_demo_gender["Percentage of Players"] = summary_age_demo_gender["Percentage of Players"].map("{:.2f}%".format)
summary_age_demo_gender

Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [15]:
# Purchase count per Age Group
purchase_count_ag = age_group_df["Purchase ID"].count()

# Average Purchase Price by Age Group 
avg_price_ag = age_group_df["Price"].mean()

# Total Purchase Value by Age Group
total_purchase_ag = age_group_df["Price"].sum()

# Average Total Purchase per Person by Age Group 
avg_purchase_ag = total_purchase_ag / age_count

In [16]:
# Create summary table for age demographics by gender 
summary_age_demo_purchase = pd.DataFrame({
    "Purchase Count": purchase_count_ag,
    "Average Purchase Price": avg_price_ag,
    "Total Purchase Value": total_purchase_ag,
    "Avg Total Purchase Value per Person": avg_purchase_ag
    
})


In [17]:
# Formatting 
summary_age_demo_purchase["Average Purchase Price"] = summary_age_demo_purchase["Average Purchase Price"].map("${:.2f}".format)
summary_age_demo_purchase["Total Purchase Value"] = summary_age_demo_purchase["Total Purchase Value"].map("${:,.2f}".format)
summary_age_demo_purchase["Avg Total Purchase Value per Person"] = summary_age_demo_purchase["Avg Total Purchase Value per Person"].map("${:.2f}".format)
summary_age_demo_purchase

Purchase Count Average Purchase Price Total Purchase Value  \
Age Group                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Avg Total Purchase Value per Person  
Age Group                                      
<10                                     $4.54  
10-14                                   $3.76  
15-19                                   $3.86  
20-24                                   $4.32  
25-29                                   $3.81  
30-34                                   $4.12  
35-39                                   $4.76  
40+                                     $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [18]:
# Group by "SN" to determine total spending per gamer 
gamer_spending_df = purchase_df.groupby(["SN"])

# Calculate purchase count by gamer 
gamer_purchase_count = gamer_spending_df["Purchase ID"].count()

# Calculate average purchase price for each gamer 
gamer_avg_price = gamer_spending_df["Price"].mean()

# Calculate total purchase value for each gamer 
gamer_total_value = gamer_spending_df["Price"].sum()

# Create dataframe that contains these values 
summary_top_spender = pd.DataFrame({
    "Purchase Count": gamer_purchase_count,
    "Average Purchase Price": gamer_avg_price,
    "Total Purchase Value": gamer_total_value,
    
})


# Sort values to determine top 5 spenders according to Total Purchase Value
top_spender_df = summary_top_spender.sort_values("Total Purchase Value", ascending=False)


In [19]:
# Formatting 
top_spender_df["Average Purchase Price"] = top_spender_df["Average Purchase Price"].map("${:.2f}".format)
top_spender_df["Total Purchase Value"] = top_spender_df["Total Purchase Value"].map("${:.2f}".format)
top_spender_df.head(5)

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [20]:
# Groupby Item ID and Item Name 
popular_items_df = purchase_df.groupby(["Item ID", "Item Name"])

# Calculate Purchase Count by Item ID 
id_purchase_count = popular_items_df["Purchase ID"].count()

# Calculate Total Purchase Value 
id_total_value = popular_items_df["Price"].sum()

# Calculate Item Price 
id_price = id_total_value / id_purchase_count

# Create dataframe that contains these values 
summary_pop_item = pd.DataFrame({
    "Purchase Count": id_purchase_count,
    "Item Price": id_price,
    "Total Purchase Value": id_total_value
    
})

# Sort values to determine top 5 spenders according to Purchase Count 
pop_df = summary_pop_item.sort_values("Purchase Count", ascending=False)


In [21]:
# Formatting 
pop_df["Item Price"] = pop_df["Item Price"].map("${:.2f}".format)
pop_df["Total Purchase Value"] = pop_df["Total Purchase Value"].map("${:.2f}".format)
pop_df.head(5)

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [22]:
# Sort values by Total Purchase Value to find the most profitable items 
profit_df = summary_pop_item.sort_values("Total Purchase Value", ascending=False)


In [23]:
# Formatting 
profit_df["Item Price"] = profit_df["Item Price"].map("${:.2f}".format)
profit_df["Total Purchase Value"] = profit_df["Total Purchase Value"].map("${:.2f}".format)
profit_df.head(5)

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Observable Trends

* 1. It is evident from the gender demographics that Heroes of Pymoli is more popular with Males where 84% of the players are Male and only 14% are Female. Males also account for approximately 84% of all purchases within the game. 

* 2. In reference to Age Demographics, the majority of players are between 20 and 24 years old, accounting for 44.8% of all players in the game. This is followed by 15 to 19 year olds that account for 18.6%. The 20-24 year olds are responsible for the largest total purchase value, at $1,114.06. However, it is the 35-39 year olds that have the highest average total purchase per individual.

* 3. The most popular and profitable item in the game is the Final Critic with over 13 purchases, amounting to $59.99 in total purchase value. 